src/Interpreters/ColumnAliasesVisitor.h

```diff
@@ -0,0 +1,76 @@
+#pragma once
+
+#include <Core/Names.h>
+#include <Interpreters/InDepthNodeVisitor.h>
+#include <Storages/ColumnsDescription.h>
+
+namespace DB
+{
+
+class IAST;
+using ASTPtr = std::shared_ptr<IAST>;
+class IDataType;
+class ASTFunction;
+class ASTIdentifier;
+using DataTypePtr = std::shared_ptr<const IDataType>;
+
+/// Visits AST node to rewrite alias columns in query
+/// Currently works only 3 kind ways below
+
+/// For example:
+//    CREATE TABLE test_table
+//    (
+//     `timestamp` DateTime,
+//     `value` UInt64,
+//     `day` Date ALIAS toDate(timestamp),
+//     `day1` Date ALIAS day + 1,
+//     `day2` Date ALIAS day1 + 1,
+//     `time` DateTime ALIAS timestamp
+//    )ENGINE = MergeTree
+//    PARTITION BY toYYYYMMDD(timestamp)
+//    ORDER BY timestamp SETTINGS index_granularity = 1;
+
+/// 1. Rewrite the filters in query when enable optimize_alias_column_prediction
+///  this could help with `optimize_trivial_count`, Partition Prune in `KeyCondition` and secondary indexes.
+///  eg: select max(value) from  test_table where day2 = today(), filters will be: ((toDate(timestamp) + 1) + 1) = today() .
+
+/// 2. Alias on alias for `required_columns` extracted in `InterpreterSelectQuery.cpp`, it could help get all dependent physical columns for query.
+///  eg: select day2 from test_table.  `required_columns` can got require columns from the tempory rewrited AST `((toDate(timestamp) + 1) + 1)`.
+
+/// 3. Help with `optimize_aggregation_in_order` and `optimize_read_in_order` in `ReadInOrderOptimizer.cpp`:
+///    For queries with alias columns in `orderBy` and `groupBy`, these ASTs will not change.
+///    But we generate tempory asts and generate tempory Actions to get the `InputOrderInfo`
+///  eg: select day1 from test_table order by day1;
+
+
+class ColumnAliasesMatcher
+{
+public:
+    using Visitor = InDepthNodeVisitor<ColumnAliasesMatcher, false>;
+
+    struct Data
+    {
+        const ColumnsDescription & columns;
+        const NameSet & forbidden_columns;
```

Comment:
It will be better to add a comment here too (I guess it can be copied from src/Storages/ReadInOrderOptimizer.cpp in some form, or just contains a reference to it)